server/controllers/stock/reports/stock/exit_report.js

```diff
@@ -0,0 +1,166 @@
+   * - exitToService : stock exit to service data
+   * - exitToDepot : stock exit to depot data
+   * - exitToLoss : stock exit as loss
+   */
+  try {
+    data = collect(params);
+  } catch (e) {
+    return next(e);
+  }
+
+  return fetchDepotDetails(params.depotUuid)
+    .then(depot => {
+      params.depotName = depot.text;
+      return data;
+    })
+    .then(groupCollection)
+    .then((bundle) => {
+      _.extend(bundle, params);
+
+      return report.render(bundle);
+    })
+    .then((result) => {
+      res.set(result.headers).send(result.report);
+    })
+    .catch(next)
+    .done();
+}
+
+/**
+ * fetchDepotDetails
+ * @param {number} depotUuid depot uuid
+ */
+function fetchDepotDetails(depotUuid) {
+  const query = 'SELECT text FROM depot WHERE uuid = ?';
+  return db.one(query, [db.bid(depotUuid)]);
+}
+
+/**
+ * @function groupCollection
+ * @description group collected data by inventory
+ */
+function groupCollection(exitCollection) {
+  const collection = {};
+
+  // exit to patient
+  collection.exitToPatient = _.groupBy(exitCollection.exitToPatient, row => row.text);
+  collection.exitToPatient = _.map(collection.exitToPatient, formatExit);
+  collection.exitToPatientEmpty = (Object.keys(collection.exitToPatient).length === 0);
+
+  // exit to service
+  collection.exitToService = _.groupBy(exitCollection.exitToService, row => row.text);
+  collection.exitToService = _.map(collection.exitToService, formatExit);
+  collection.exitToServiceEmpty = (Object.keys(collection.exitToService).length === 0);
+
+  // exit to depot
+  collection.exitToDepot = _.groupBy(exitCollection.exitToDepot, row => row.text);
+  collection.exitToDepot = _.map(collection.exitToDepot, formatExit);
+  collection.exitToDepotEmpty = (Object.keys(collection.exitToDepot).length === 0);
+
+  // exit to loss
+  collection.exitToLoss = _.groupBy(exitCollection.exitToLoss, row => row.text);
+  collection.exitToLoss = _.map(collection.exitToLoss, formatExit);
+  collection.exitToLossEmpty = (Object.keys(collection.exitToLoss).length === 0);
+
+  return collection;
+}
+
+/**
+ * @function formatExit
+ */
+function formatExit(value, index) {
+  return {
+    inventory_name : index,
+    inventory_unit : value && value[0] ? value[0].unit_text : '',
+    inventory_stock_exit_data : value,
+    inventory_stock_exit_quantity : _.sumBy(value, 'quantity'),
+  };
+}
+
+/**
+ * @function collect
+ * @param {object} params query parameters
+ * @return {object} { exitToPatient: [], exitToService: [], exitToDepot: [], exitToLoss: [] }
+ */
+async function collect(params) {
+  const {
+    depotUuid,
+    dateFrom,
+    dateTo,
+    showDetails,
+    includePatientExit,
+    includeServiceExit,
+    includeDepotExit,
+    includeLossExit,
+  } = params;
+
+  const data = {};
+
+  try {
+    // get stock exit to patient
+    if (includePatientExit) {
+      data.exitToPatient = await StockExitToPatient.fetch(depotUuid, dateFrom, dateTo, showDetails);
```

Comment:
This is cool!  However, it will also be kind of slow.  It says that we will wait for each one of these values in series.

You can speed it up by executing these queries in parallel.  Just push each value into an array and use `Promise.all()`!  Like this:
```js
const promises = [];
if (includePatientExit) {
  promises.push(StockExitToPatient.fetch(depotUuid, dateFrom, dateTo, showDetails))
}

 // etc, do that for the rest..

 return q.all(promises)

   // convert from array to data object
  .spread(exitToPatient, exitToService /* .. etc */) => ({
    exitToPatient,
    exitToService,
    // etc...
  }));
```
[Here's a fun link that shows cool things you can do with promises](http://exploringjs.com/es6/ch_promises.html#_forking-and-joining-computations-via-promiseall).  Also, [this SO answer](https://stackoverflow.com/a/31414472) gives some really good rules about how to think about promises.